### PR TITLE
moving tags to an array field in aspect model + data migration for tags

### PR DESCRIPTION
**Pull request by shriramshankar**
tags refactoring for aspects along with data migration for subject and aspect tags.

### CODE REVIEW

File: api/v1/swagger.yaml [hunk 1]
```diff
@@ -343,9 +343,12 @@ paths:
               tags:
                 type: array
                 items:
-                  $ref: "#/definitions/TagRequest"
+                  type: string
+                  maxLength: 60
+                  pattern: ^[0-9A-Za-z_][0-9A-Za-z_\\-]{1,60}$
```
Comment (iamigo) [inline api/v1/swagger.yaml]: I think this should be {0,59} now

File: api/v1/swagger.yaml [hunk 2]
```diff
@@ -569,9 +572,12 @@ paths:
...
```
Comment (iamigo) [inline api/v1/swagger.yaml]: I think this should be {0,59} now

File: api/v1/swagger.yaml [hunk 3]
```diff
@@ -728,9 +734,12 @@ paths:
...
```
Comment (iamigo) [inline api/v1/swagger.yaml]: I think this should be {0,59} now

File: api/v1/swagger.yaml [hunk 4]
```diff
@@ -3162,10 +3171,10 @@ paths:
                 items:
                   type: string
                   maxLength: 60
-                  pattern: ^[0-9A-Za-z_\\-]{1,60}$
+                  pattern: ^[0-9A-Za-z_][0-9A-Za-z_\\-]{1,60}$
```
Comment (iamigo) [inline api/v1/swagger.yaml]: I think this should be {0,59} now

File: api/v1/swagger.yaml [hunk 5]
```diff
@@ -3331,10 +3340,10 @@ paths:
...
```
Comment (iamigo) [inline api/v1/swagger.yaml]: I think this should be {0,59} now

File: api/v1/swagger.yaml [hunk 6]
```diff
@@ -3436,10 +3445,10 @@ paths:
...
```
Comment (iamigo) [inline api/v1/swagger.yaml]: I think this should be {0,59} now

File: api/v1/swagger.yaml [hunk 7]
```diff
@@ -3541,10 +3550,10 @@ paths:
...
```
Comment (iamigo) [inline api/v1/swagger.yaml]: I think this should be {0,59} now

File: api/v1/swagger.yaml [hunk 8]
```diff
@@ -4835,12 +4803,14 @@ definitions:
         description: >
           TODO
       tags:
-        readOnly: true
         type: array
         items:
-          $ref: "#/definitions/TagResponse"
+          type: string
+          maxLength: 60
+          pattern: ^[0-9A-Za-z_][0-9A-Za-z_\\-]{1,60}$
```
Comment (iamigo) [inline api/v1/swagger.yaml]: I think this should be {0,59} now

File: api/v1/swagger.yaml [hunk 9]
```diff
@@ -5416,10 +5386,10 @@ definitions:
         items:
           type: string
           maxLength: 60
-          pattern: ^[0-9A-Za-z_\\-]{1,60}$
+          pattern: ^[0-9A-Za-z_][0-9A-Za-z_\\-]{1,60}$
```
Comment (iamigo) [inline api/v1/swagger.yaml]: I think this should be {0,59} now

File: db/model/aspect.js [hunk 10]
```diff
@@ -110,6 +110,11 @@ module.exports = function aspect(seq, dataTypes) {
         },
       },
     },
+    tags: {
```
Comment (iamigo) [inline db/model/aspect.js]: validation?

Comment (shriramshankar) [inline db/model/aspect.js]: All the validations are in the API layer. Doing it in the db layer will be redundant, I guess.

Comment (shriramshankar) [inline db/model/aspect.js]: Let me know if you want me to add it in the db layer too.

**Comment by shriramshankar** [comment]
@iamigo PR updated with the swagger changes.

**Comment by pallavi2209** [comment]
I think the admin front end tag (get and post) related things needs to be updated with this change. Do we handle that separately?

**Comment by shriramshankar** [comment]
@iamigo  @pallavi2209 
1. filters work
2. perspective loads
3. hierarchy loads about 40% faster
there is something wrong with migration, the tags data get migrated correctly, but it looks like tags wit h '_' in the name get copied multiple times.
http://refocus-pr-25.herokuapp.com/v1/subjects/Salesforce.SFDC_Core.CHI.SP1.CS10    I tweaked the migrating code a couple of times but do not see a better result. Any pointers?
I created a another branch to play with the migration code,
https://github.com/salesforce/refocus/blob/b801984561a5a3655017a8ad62470ccaf2f87d40/migrations/20161002203022-moveTagDataToField.js

**Comment by iamigo** [comment]
Any progress on the data migration? If you want to go over this together on a hangout let me know.

**Comment by shriramshankar** [comment]
@iamigo I have another idea that I want to try. Have trust1 meeting for ssl till 11.30am. Will try it after that and let you know.

**Comment by shriramshankar** [comment]
@iamigo updated the pr with migrations fixes.

**Comment by iamigo** [comment]
OK to merge.

**Comment by shriramshankar** [comment]
@pallavi2209 can you merge it?

**Pull request merged**